server/replication/replication_mode.go

```diff
@@ -396,78 +411,158 @@ func (m *ModeManager) tickDR() {
+	stores := m.checkStoreStatus()
 
 	// canSync is true when every region has at least 1 replica in each DC.
-	canSync := downPrimaryStores < totalPrimaryPeers && downDrStores < totalDrPeers &&
-		upPrimayStores > 0 && upDrStores > 0
+	canSync := len(stores[primaryDown]) < totalPrimaryPeers && len(stores[drDown]) < totalDrPeers &&
+		len(stores[primaryUp]) > 0 && len(stores[drUp]) > 0
 
 	// hasMajority is true when every region has majority peer online.
 	var upPeers int
-	if downPrimaryStores < totalPrimaryPeers {
-		upPeers += totalPrimaryPeers - downPrimaryStores
+	if len(stores[primaryDown]) < totalPrimaryPeers {
+		upPeers += totalPrimaryPeers - len(stores[primaryDown])
 	}
-	if downDrStores < totalDrPeers {
-		upPeers += totalDrPeers - downDrStores
+	if len(stores[drDown]) < totalDrPeers {
+		upPeers += totalDrPeers - len(stores[drDown])
 	}
 	hasMajority := upPeers*2 > totalPrimaryPeers+totalDrPeers
 
 	log.Debug("replication store status",
-		zap.Int("up-primary", upPrimayStores),
-		zap.Int("up-dr", upDrStores),
-		zap.Int("down-primary", downPrimaryStores),
-		zap.Int("down-dr", downDrStores),
+		zap.Uint64s("up-primary", stores[primaryUp]),
+		zap.Uint64s("up-dr", stores[drUp]),
+		zap.Uint64s("down-primary", stores[primaryDown]),
+		zap.Uint64s("down-dr", stores[drDown]),
 		zap.Bool("can-sync", canSync),
 		zap.Int("up-peers", upPeers),
 		zap.Bool("has-majority", hasMajority),
 	)
 
-	// If hasMajority is false, the cluster is always unavailable. Switch to async won't help.
-	if !canSync && hasMajority && m.drGetState() != drStateAsync && m.drCheckAsyncTimeout() {
-		m.drSwitchToAsync()
-	}
-
-	if canSync && m.drGetState() == drStateAsync {
-		m.drSwitchToSyncRecover()
-	}
-
-	if m.drGetState() == drStateSyncRecover {
-		m.updateProgress()
-		progress := m.estimateProgress()
-		drRecoverProgressGauge.Set(float64(progress))
-
-		if progress == 1.0 {
+	/*
+
+	           +----+      all region sync     +------------+
+	           |SYNC| <----------------------- |SYNC_RECOVER|
+	           +----+                          +------------+
+	             |^                                 ^ |
+	      DR down||                                 | |DR down
+	             ||DR up                       DR up| |
+	             v|                                 | v
+	        +----------+     all tikv sync        +-----+
+	   +--> |ASYNC_WAIT|------------------------> |ASYNC|<------+
+	   |    +----------+                          +-----+       |
+	   |            |                              |            |
+	   |tikv up/down|                              |tikv up/down|
+	   +------------+                              +------------+
+
+	*/
+
+	switch m.drGetState() {
+	case drStateSync:
+		// If hasMajority is false, the cluster is always unavailable. Switch to async won't help.
+		if !canSync && hasMajority && m.drCheckAsyncTimeout() {
+			m.drSwitchToAsyncWait(stores[primaryUp])
+		}
+	case drStateAsyncWait:
+		if canSync {
```

Comment:
Can this directly switch to `sync`? may the follower replica have a large logging lag during `AsyncWait`?

Reply:
It's ok. In `AsyncWait` state, TiKV cannot write data so there will be no log lag.

Reply:
`Sync` -> `AsyncWait` already wait for a timeout, why cannot write data? In my understanding, once switch to `AsyncWait`, it means there are majority replicas in primary DC, but cannot replicate to DR DC's replica.

Reply:
| State  | Can commit raftlog locally | Transport between DC |
| ------------- | ------------- | --------------- |
| Sync  | No | Connected |
| AsyncWait  | No  | Disconnected |
| Async | Yes | Disconnected |

Without the `AsyncWait` state, it will appear that some tikv are in Sync state and some TiKV are in Async state. Because TiKV in Async state can commit successfully with only 1DC participating in the replication log, the ACID of the dr dc can be broken.
From the table we can see in `AsyncWait` state, the network is cut off, and it cannot commit raftlog locally, so it in fact cannot commit raft log successfully.

Reply:
got